bundles/core/src/main/java/com/adobe/cq/wcm/core/components/internal/models/v4/ListImpl.java

```diff
@@ -0,0 +1,133 @@
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~*/
+package com.adobe.cq.wcm.core.components.internal.models.v4;
+
+import com.adobe.cq.export.json.ComponentExporter;
+import com.adobe.cq.export.json.ExporterConstants;
+import com.adobe.cq.wcm.core.components.commons.link.Link;
+import com.adobe.cq.wcm.core.components.commons.link.LinkManager;
+import com.adobe.cq.wcm.core.components.internal.link.LinkManagerImpl;
+import com.adobe.cq.wcm.core.components.internal.models.v1.AbstractListItemImpl;
+import com.adobe.cq.wcm.core.components.models.List;
+import com.adobe.cq.wcm.core.components.models.ListItem;
+import com.day.cq.wcm.api.Page;
+import com.day.cq.wcm.api.components.Component;
+import org.apache.commons.lang3.ObjectUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.sling.api.SlingHttpServletRequest;
+import org.apache.sling.api.resource.Resource;
+import org.apache.sling.models.annotations.Exporter;
+import org.apache.sling.models.annotations.Model;
+import org.jetbrains.annotations.NotNull;
+
+import javax.annotation.PostConstruct;
+import java.text.Collator;
+import java.util.Collection;
+import java.util.Comparator;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+@Model(adaptables = SlingHttpServletRequest.class, adapters = {List.class, ComponentExporter.class}, resourceType = ListImpl.RESOURCE_TYPE)
+@Exporter(name = ExporterConstants.SLING_MODEL_EXPORTER_NAME, extensions = ExporterConstants.SLING_MODEL_EXTENSION)
+public class ListImpl extends com.adobe.cq.wcm.core.components.internal.models.v3.ListImpl implements List {
+
+    protected static final String RESOURCE_TYPE = "core/wcm/components/list/v4/list";
+
+    protected ListItem newPageListItem(@NotNull LinkManager linkManager, @NotNull Page page, String parentId, Component component) {
+        return new PageListItemImpl(linkManager.get(page).build(), page, parentId, component, showDescription, linkItems || displayItemAsTeaser, resource);
+    }
+
+    /**
+     * Initialize the model.
+     */
+    @PostConstruct
+    protected void initModel() {
+        super.initModel();
+        if (Source.STATIC.equals(getListType())) {
+            if (this.listItems == null) {
+                Resource staticNode = this.resource.getChild(NN_STATIC);
+                if (staticNode != null) {
+                    this.listItems = getStaticListItems();
+                }
+            }
+        }
+    }
+
+    private Collection<ListItem> getStaticListItems() {
+        AtomicBoolean hasExternalLink = new AtomicBoolean(false);
```

Comment:
Why do we need a AtomicBoolean in this case?

Reply:
because the local variable must be `final` to be accessible in the lambda, where it is mutated. 

Maybe using a regular for-each loop to create the list of items would be more readable here and would not require the atomic boolean.